lib/kubernetes-deploy/resource_watcher.rb

```diff
@@ -17,19 +17,16 @@ def initialize(resources:, sync_mediator:, logger:, deploy_started_at: Time.now.
     end
 
     def run(delay_sync: 3.seconds, reminder_interval: 30.seconds, record_summary: true)
-      delay_sync_until = last_message_logged_at = monitoring_started = Time.now.utc
+      last_message_logged_at = monitoring_started = Time.now.utc
       remainder = @resources.dup
 
       while remainder.present?
-        if @timeout && (Time.now.utc - monitoring_started > @timeout)
-          report_and_give_up(remainder)
-        end
-        if (sleep_duration = delay_sync_until - Time.now.utc) > 0
-          sleep(sleep_duration)
-        end
-        delay_sync_until = Time.now.utc + delay_sync # don't pummel the API if the sync is fast
+        report_and_give_up(remainder) if global_timeout?(monitoring_started)
+        sleep_until_next_sync(delay_sync)
 
         @sync_mediator.sync(remainder)
+        remainder.each(&:post_sync)
```

Comment:
As implemented `sync` is run all on all the resources then `post_sync` is called. In rails /  https://api.rubyonrails.org/classes/ActiveSupport/Callbacks.html `post_sync` would trigger immediately after each resource has `sync` called. Is the implemented behavior needed? If not I think it would be preferable to do it the more traditional way.

Reply:
tl;dr The technical reason is that `sync` runs concurrently and `post_sync` shouldn't. I also don't see the value of the indirection (vs writing the code at the end of `sync` itself) if it were executed immediately after.

The idea here is that `sync` should exclusively collect state from the API server and should not have any side-effects. It is run in parallel, so weird things can happen if the instance associates additional behaviours with it (e.g. seems to me that if two pods complete in different sync groups during the same polling cycle today, their logs will get interleaved). The `post_sync` hook (naming suggestions appreciated) is a way to let instances do something every polling cycle if they need to, with the guarantee that (a) their instance data is already up to date; and (b) nothing else is logging at the same time. I considered making this `dump_logs` or something, but thought it would be better design to keep the watcher agnostic about the details of what instances might want to do.

Reply:
(b) seems like its just a result of running the hook serially. If the goal is to keep the watcher agnostic, which I agree with,  I'm think a more descriptive name would help `serialized_post_sync` (I'm terrible at naming :( )